Added transformAsync ability to modules

Dr15Jones opened the pull request:
#### PR description:

This is similar to transform except two actions are specified similar to acquire/produce.

#### PR validation:

Code compiles and all framework unit tests pass.

Dr15Jones committed:
Added transformAsync ability to modules

This is similar to transform except two actions are specified similar to acquire/produce.

Dr15Jones commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39557/32347

- This PR adds an extra 92KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/interface/global/EDFilterBase.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/global/EDProducerBase.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/global/implementors.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/limited/EDFilterBase.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/limited/EDProducerBase.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/limited/implementors.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/one/EDFilterBase.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/one/EDProducerBase.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/one/implementors.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/stream/ProducingModuleAdaptorBase.h  modified in PR(s): #38801
  - File  FWCore/Framework/src/global/EDFilterBase.cc  modified in PR(s): #38801
  - File  FWCore/Framework/src/global/EDProducerBase.cc  modified in PR(s): #38801
  - File  FWCore/Framework/src/limited/EDFilterBase.cc  modified in PR(s): #38801
  - File  FWCore/Framework/src/limited/EDProducerBase.cc  modified in PR(s): #38801
  - File  FWCore/Framework/src/one/EDFilterBase.cc  modified in PR(s): #38801
  - File  FWCore/Framework/src/one/EDProducerBase.cc  modified in PR(s): #38801
  - File  FWCore/Framework/src/stream/ProducingModuleAdaptorBase.cc  modified in PR(s): #38801
  - File  FWCore/Framework/test/global_filter_t.cppunit.cc  modified in PR(s): #38801
  - File  FWCore/Framework/test/global_producer_t.cppunit.cc  modified in PR(s): #38801
  - File  FWCore/Integration/test/BuildFile.xml  modified in PR(s): #38801

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

- FWCore/Framework (**core**)
- FWCore/Integration (**core**)


@makortel, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@makortel, @missirol, @wddgit this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-53fe2e/27890/summary.html
**COMMIT**: 62340ccbfc218026780ff0ea9007b0c0355a250e
**CMSSW**: CMSSW_12_6_X_2022-09-30-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39557/27890/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

Dr15Jones commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39557/32351

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/interface/global/EDFilterBase.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/global/EDProducerBase.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/global/implementors.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/limited/EDFilterBase.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/limited/EDProducerBase.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/limited/implementors.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/one/EDFilterBase.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/one/EDProducerBase.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/one/implementors.h  modified in PR(s): #38801
  - File  FWCore/Framework/interface/stream/ProducingModuleAdaptorBase.h  modified in PR(s): #38801
  - File  FWCore/Framework/src/global/EDFilterBase.cc  modified in PR(s): #38801
  - File  FWCore/Framework/src/global/EDProducerBase.cc  modified in PR(s): #38801
  - File  FWCore/Framework/src/limited/EDFilterBase.cc  modified in PR(s): #38801
  - File  FWCore/Framework/src/limited/EDProducerBase.cc  modified in PR(s): #38801
  - File  FWCore/Framework/src/one/EDFilterBase.cc  modified in PR(s): #38801
  - File  FWCore/Framework/src/one/EDProducerBase.cc  modified in PR(s): #38801
  - File  FWCore/Framework/src/stream/ProducingModuleAdaptorBase.cc  modified in PR(s): #38801
  - File  FWCore/Framework/test/global_filter_t.cppunit.cc  modified in PR(s): #38801
  - File  FWCore/Framework/test/global_producer_t.cppunit.cc  modified in PR(s): #38801
  - File  FWCore/Integration/test/BuildFile.xml  modified in PR(s): #38801

cmsbuild commented:
Pull request #39557 was updated. @makortel, @smuzaffar, @Dr15Jones can you please check and sign again.

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-53fe2e/27898/summary.html
**COMMIT**: 1cf94616adb2d64da4432aaa748e48c0fc8f1e56
**CMSSW**: CMSSW_12_6_X_2022-09-30-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39557/27898/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test createDBObjecs had ERRORS
---> test GeometryTrackerGeometryBuilderTestDriver had ERRORS
---> test testPhase2SkimmedGeometry had ERRORS
---> test testPhase2PixelNtuple had ERRORS
and more ...
</pre>

## RelVals-INPUT

The relvals timed out after 4 hours.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-53fe2e/41834.0_TTbar_14TeV+2026D94+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 3174 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3432650
 - DQMHistoTests: Total failures: 8230
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3424398
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 204 log files, 49 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-09-30-2300+53fe2e/53091/triggerResults): no differences found

Dr15Jones commented:
Please test

makortel commented on FWCore/Framework/src/WorkerT.cc:
```diff
@@ -269,63 +269,84 @@ namespace edm {
+                                               EventPrincipal const& iEvent,
+                                               ParentContext const& iParent,
+                                               ServiceWeakToken const& weakToken) {
+    try {
```
```suggestion
    CMS_SA_ALLOW try {
```

makortel commented on FWCore/Framework/src/TransformerBase.cc:
```diff
@@ -48,15 +61,51 @@ namespace edm {
+        return;
+      }
+      if (handle->wrapper()) {
+        auto cache = std::make_shared<std::any>();
```
I _believe_ in principle the `cache` could be also `std::shared_ptr<void>` (in which case the `TransformBase::PreTransformFunction` should probably return the `std::shared_ptr<void>`). I think `shared_ptr<void>` would allow supporting move-only types (assuming the copyability of `shared_ptr` itself is sufficient), and it would avoid the (possibly minuscule) run-time cost of `std::any_cast`. I _believe_ the code around is sufficient to guarantee type safety even with `void`.

makortel replied:
Nevermind, I figured it out why my suggestion above would not work. The space (object) shared between the "acquire" and "produce" parts really needs to be created here.

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-53fe2e/27930/summary.html
**COMMIT**: 1cf94616adb2d64da4432aaa748e48c0fc8f1e56
**CMSSW**: CMSSW_12_6_X_2022-10-02-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39557/27930/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test createDBObjecs had ERRORS
---> test testHeterogeneousCoreAlpakaTestWriteRead had ERRORS
---> test materialBudgetHGCalPlots had ERRORS
</pre>

## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-53fe2e/41834.0_TTbar_14TeV+2026D94+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 3171 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3432650
 - DQMHistoTests: Total failures: 8227
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3424401
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 204 log files, 49 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-10-02-2300+53fe2e/53122/triggerResults): no differences found

Dr15Jones committed:
Changes based on PR comments

- Use enum for SOA column usage
- Added CMS_SA_ALLOW to try catch blocks
- Better handling of std::any

Dr15Jones commented:
please test

cmsbuild commented:
-1

**Failed Tests**: AddOn 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-53fe2e/28083/summary.html
**COMMIT**: b40a43a04d0edda0c03f95cfb48ef956979f1637
**CMSSW**: CMSSW_12_6_X_2022-10-06-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39557/28083/install.sh` to create a dev area with all the needed externals and cmssw changes.


## AddOn Tests

- [hlt_mc_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-53fe2e/28083/addOnTests/hlt_mc_HIon)
```
----- Begin Fatal Exception 07-Oct-2022 00:23:53 CEST-----------------------
An exception of category 'NotFound' occurred while
   [0] Calling EventProcessor::runToCompletion (which does almost everything after beginJob and before endJob)
Exception Message:
Service no ServiceRegistry has been set for this thread 
----- End Fatal Exception -------------------------------------------------
```


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-53fe2e/41834.0_TTbar_14TeV+2026D94+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3391103
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3391075
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 204 log files, 49 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-10-06-1100+53fe2e/53251/triggerResults): no differences found

Dr15Jones commented:
abort

Dr15Jones commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-53fe2e/28212/summary.html
**COMMIT**: b40a43a04d0edda0c03f95cfb48ef956979f1637
**CMSSW**: CMSSW_12_6_X_2022-10-12-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39557/28212/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test test_CreateFileLists had ERRORS
---> test testSubmitters had ERRORS
---> test testAlignmentOfflineValidation had ERRORS
---> test test_DetectorStateFilter had ERRORS
and more ...
</pre>

## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-53fe2e/20834.0_TTbar_14TeV+2026D88+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-53fe2e/20834.75_TTbar_14TeV+2026D88_HLT75e33+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HLT75e33+HARVESTGlobal
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-53fe2e/20896.0_CloseByPGun_CE_E_Front_120um+2026D88+CE_E_Front_120um_GenSimHLBeamSpotHGCALCloseBy+DigiTrigger+RecoGlobal+HARVESTGlobal
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-53fe2e/20900.0_CloseByPGun_CE_H_Coarse_Scint+2026D88+CE_H_Coarse_Scint_GenSimHLBeamSpotHGCALCloseBy+DigiTrigger+RecoGlobal+HARVESTGlobal
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-53fe2e/21034.999_TTbar_14TeV+2026D88PU_PMXS1S2PR+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+PREMIX_PremixHLBeamSpot14PU+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-53fe2e/23234.0_TTbar_14TeV+2026D94+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 7 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3392309
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3392278
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 204 log files, 49 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-10-12-1100+53fe2e/53361/triggerResults): no differences found

Dr15Jones commented:
@smuzaffar @makortel the unit tests failure are either
- voms problems or
- xrtood problems

Is there possibly an issue with one of the testing machines?

smuzaffar commented:
please test

Yes it could be grid certificate validation issue as tests were run on a [node](https://cmssdt.cern.ch/jenkins/job/ib-run-pr-tests/28212/) which had cvmfs issues and `/cvmfs/grid.cern.ch` was unmounted on it. Lets re-run it

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39557/32567

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/WorkerT.cc  modified in PR(s): #39710

makortel commented:
@cmsbuild, please abort

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-53fe2e/28243/summary.html
**COMMIT**: 8924cf4ac43b324ccfebb432fc20b16121efbd7f
**CMSSW**: CMSSW_12_6_X_2022-10-13-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39557/28243/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 7 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3392309
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3392278
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 204 log files, 49 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-10-13-1100+53fe2e/53389/triggerResults): no differences found

makortel commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

rappoccio commented:
@Dr15Jones can you give a bit more explanation here? It's a relatively in-depth PR that I'm not fully understanding. Was there a talk or something we can look at?

makortel commented:
@rappoccio The Transformer ability was added in https://github.com/cms-sw/cmssw/pull/38454 with a synchronous interface. Addition of an asynchronous interface was discussed in https://github.com/cms-sw/cmssw/pull/38454#issuecomment-1191784638 (and onwards).

The Transformer ability is used in https://github.com/cms-sw/cmssw/pull/39428 for implicit/automated data product transfers from device memory space to host memory space. After we get this PR merged, I'll make use of the asynchronous interface in #39428 to make the device-to-host transfers asynchronous.

rappoccio commented:
+1
